feat(pxe): resolve tagging secret strategy via wallet hook (#24040)
AztecProtocol/aztec-packages: docs/docs-developers/docs/foundational-topics/pxe/execution_hooks.md

---
title: Execution hooks
sidebar_position: 3
tags: [pxe, wallets]
description: How wallets use PXE execution hooks to apply custom policies during client-side simulation.
---

> Execution hooks are callbacks that the PXE invokes during client-side simulation when an operation needs a decision from the wallet. They let the wallet apply its own policies before execution proceeds, such as prompting the user, consulting a dynamic allowlist, or inspecting call arguments. Every hook is optional, and when a hook is absent the PXE applies a safe default.

Reviewer: ```suggestion
Execution hooks are callbacks that the PXE invokes during client-side simulation when an operation needs a decision from the wallet. They let the wallet apply its own policies before execution proceeds, such as prompting the user, consulting a dynamic allowlist, or inspecting call arguments. All hooks are optional, and when a hook is absent the PXE applies a safe default.
```

## Configuring hooks

Pass a `hooks` object when creating the PXE:

```typescript
import { createPXE } from "@aztec/pxe/server";
import { DeliveryPrivacyPreference } from "@aztec/pxe/config";

const pxe = await createPXE(node, config, {
  hooks: {
    // Allow calls to a known helper contract, deny everything else.
    authorizeUtilityCall: async (request) => {
      return request.target.equals(trustedHelper)
        ? { authorized: true }
        : { authorized: false, reason: "Unknown target" };
    },
    // Accept the privacy leak of on-the-fly handshakes so messages reach recipients that haven't registered the sender.

>     getDeliveryPrivacyPreference: async () => DeliveryPrivacyPreference.BEST_EFFORT,

Reviewer: Re. naming, see how the comment is completely disconnected from any notion of 'best effort'.

Author: Will mark this as resolved, and we can continue the conversation on the other comment thread. But I agree

  },
});
```

## `authorizeUtilityCall`

> Called whenever a utility function makes a cross-contract call. A call made by a malicious contract could leak private information, so the hook lets the wallet decide, per call, whether to allow it. A static allowlist would not work here because neither the app nor the wallet can predict ahead of time which contracts will be invoked during execution: permission must be asked after execution has begun. Calls to standard contracts (such as the HandshakeRegistry, which is queried during every contract's sync) bypass this hook and are always authorized.

Reviewer: I think we need a concrete example showing what the danger is, and why in some circumstances this is ok. Otherwise this is too vague, the reader is not given sufficient context and information to make a decision on how to allow calls.

> Unlike [authentication witnesses (authwits)](../../aztec-js/how_to_use_authwit.md), the hook is invoked live, while execution is underway. Authwits can be recorded during simulation and signed once at the end, but the PXE cannot predict what a utility call would return, so it must ask before continuing. Most of the time the wallet can answer on its own, for example against a list of audited or previously trusted contracts, without involving the user.

Reviewer: >Most of the time the wallet can answer on its own

Not only can, it probably _should_, in order to avoid interrupting execution multiple times asking the user for confirmation.

> ### In production

Reviewer: I'd swap prod and tests so that tests are before

Pass an `authorizeUtilityCall` hook when [creating the PXE](#configuring-hooks). It receives a `UtilityCallAuthorizationRequest` with the caller and target addresses, their contract class IDs, the function selector, the function name, the arguments, and the caller context (`'private'`, `'private view'`, or `'utility'`). Return `{ authorized: true }` to allow the call, or `{ authorized: false, reason: '...' }` to deny it with a message.

When the hook is absent, cross-contract utility calls are denied. See [Cross-contract utility call denied](../../aztec-nr/debugging.md#cross-contract-utility-call-denied) for the resulting error.

### In Noir tests

> The hook only exists on a real PXE. When testing cross-contract utility calls in the Noir test environment (TXE), use `with_authorized_utility_call_targets` on your call options:

Reviewer: We don't talk about TXE, TXE is an implementation detail of `aztec test`. I'd specially avoid talking about 'real PXEs'.

```suggestion
When testing cross-contract utility calls in Noir using `TestEnvironment`, use `with_authorized_utility_call_targets` on your call options:
```

```rust
// For private calls:
env.call_private_opts(
    account,

>     CallPrivateOptions::new().with_authorized_utility_call_targets([target_address]),

Reviewer: Might be nice to have an 'allow all' option in `TestEnv::new`

Author: Will create a follow up issue for it

Author: Filed as F-733: https://linear.app/aztec-labs/issue/F-733/add-an-authorize-all-utility-call-targets-option-to-the-txe

    ExecuteUtilityOptions::new().with_authorized_utility_call_targets([target_address]),
    MyContract::at(caller).some_utility_fn(),
);
```

## `getDeliveryPrivacyPreference`

Called when message delivery needs a tagging secret and the executing contract has not pinned a tag-secret derivation. The hook lets the wallet choose between maximum privacy and delivery that requires no sender-recipient coordination; see [Delivery privacy preference](../../aztec-nr/framework-description/note_delivery.md#delivery-privacy-preference) for the trade-offs and the defaults in each environment.

### In production

Pass a `getDeliveryPrivacyPreference` hook when [creating the PXE](#configuring-hooks). It receives a `DeliveryPrivacyPreferenceRequest` with the executing contract's address and the message's sender, recipient, and delivery mode (`'constrained'` or `'unconstrained'`), so a wallet can apply per-application or per-recipient policies, or surface the decision to the user, instead of returning a fixed value.

When the hook is absent, the PXE assumes `DeliveryPrivacyPreference.MAX_PRIVACY`, so privacy is never weakened without the wallet opting in.

### In Noir tests

> The hook only exists on a real PXE. In the Noir test environment (TXE), which defaults to max privacy like a bare PXE, set the preference on the test environment; it affects message delivery in subsequent private executions:

Reviewer: Same comment re. txe and ordering of sections

```rust

> env.set_delivery_privacy_preference(DeliveryPrivacyPreference::best_effort());

Reviewer: Why here and not in `TestEnv::new`?

---
title: Execution hooks
sidebar_position: 3
tags: [pxe, wallets]
description: How wallets use PXE execution hooks to apply custom policies during client-side simulation.
---

> Execution hooks are callbacks that the PXE invokes during client-side simulation when an operation needs a decision from the wallet. They let the wallet apply its own policies before execution proceeds, such as prompting the user, consulting a dynamic allowlist, or inspecting call arguments. All hooks are optional, and when a hook is absent the PXE applies a safe default.

Reviewer 2: nit: default is set by the PXE initialization caller right? So it is not necessarily "safe"?

Author: No, by "default" I mean what happens when no hook is set when PXE is created. It's safe in the sense that it avoid privacy leaks (I'll make it clearer).

### In Noir tests

When testing in Noir, leaving the strategy unset makes `TestEnvironment` fall back to the bare PXE default. Set a strategy when creating the environment to exercise a specific one; it affects message delivery in private executions:

```rust
let env = TestEnvironment::new_opts(
    TestEnvironmentOptions::new().with_tagging_secret_strategy(TaggingSecretStrategy::non_interactive_handshake()),
);
```

### In production

Pass a `resolveTaggingSecretStrategy` hook when [creating the PXE](#configuring-hooks). It receives a `TaggingSecretStrategyRequest` with the executing contract's address and the message's sender, recipient, and delivery mode (`'constrained'` or `'unconstrained'`), so a wallet can apply per-application or per-recipient policies, or surface the decision to the user, instead of returning a fixed value.

> When the hook is absent, the PXE applies a privacy-safe default: unconstrained delivery uses an address-derived (Diffie-Hellman) shared secret, which leaves no onchain trace, while constrained delivery fails rather than silently revealing the recipient through a non-interactive handshake.

Reviewer 2: should we just link to the address-derived section here rather than repeating (Diffie-Hellman) and same for the handshake case
